added types for '@nick-thompson/elementary'

Hrle97 opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.


Elementary is a runtime for JS similar to NodeJS. It can run anything that can compile to JS and it exposes some global variables that need typings. The package '@nick-thompson/elementary' wraps up this global variable nicely and provides extra functionality to the runtime.

<s>
TSLint has some exceptions to make the code more readable. There is also redundant JSDoc because some function types are hard to represent with just TypeScript.
package.json contains some unwanted fields to make testing easier. ([DefinitelyTyped-tools PR](https://github.com/microsoft/DefinitelyTyped-tools/pull/314))
There is a .editorconfig in the folder to limit lines to 80 which makes development a bit easier for me. I don't know if this is allowed.
</s>

sandersn commented on types/nick-thompson__elementary/package.json:
```diff
@@ -0,0 +1,16 @@
+{
+    "private"        : true,
+    "dependencies"   : {
+        "@types/node": "*"
+    },
```
DT infers `@types/*` dependencies, so you don't have to do it yourself.

sandersn commented on types/nick-thompson__elementary/package.json:
```diff
@@ -0,0 +1,16 @@
+{
```
This file is not needed and should be removed. I'll provide additional details.

sandersn commented on types/nick-thompson__elementary/package.json:
```diff
@@ -0,0 +1,16 @@
+    "dependencies"   : {
+        "@types/node": "*"
+    },
+    "devDependencies": {
+        "@nick-thompson/elementary": "^0.9.6",
+        "typescript"               : "~4.4.2",
+        "colors"                   : "^1.4.0"
+    },
```
devDeps are not allowed in DT package.jsons. Also, I don't see any of these dependencies used in the code.

sandersn commented on types/nick-thompson__elementary/package.json:
```diff
@@ -0,0 +1,16 @@
+        "typescript"               : "~4.4.2",
+        "colors"                   : "^1.4.0"
+    },
+    "scripts"        : {
+        "build-test"    : "tsc --outDir ./build ./nick-thompson__elementary-tests.ts",
+        "test"          : "elementary -i 0 -o 2 ./build/nick-thompson__elementary-tests.js",
+        "build-and-test": "npm run build-test & npm run test"
+    }
```
DT packages should only use DT's test infrastructure in the top-level package.json.

sandersn commented:
The package.json is not needed, given the way that DT infrastructure works. The `@types` package publisher will generate a correct package.json just from the .d.ts files.

Hrle97 commented:
> The package.json is not needed, given the way that DT infrastructure works. The `@types` package publisher will generate a correct package.json just from the .d.ts files.

I removed it in recent commits, so it shouldn't be there anymore.
The commit where I commented, "fix package.json" I fixed the root package.json because I added an unnecessary dependency there by accident.

sandersn commented:
You're right, I mis-read the diff.

Hrle97 commented:
Ready to merge
